File: lightbulb-dialogs/src/main/java/com/github/RooneyAndShadows/lightbulb/dialogs/picker_dialog_time/TimePickerDialog.java
package com.github.rooneyandshadows.lightbulb.dialogs.picker_dialog_time;

import android.os.Build;
import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.View;
import android.widget.TimePicker;

import com.github.rooneyandshadows.java.commons.date.DateUtils;
import com.github.rooneyandshadows.lightbulb.dialogs.base.LightBulbPickerDialogFragment;
import com.github.rooneyandshadows.lightbulb.dialogs.R;

import java.util.Date;

@SuppressWarnings({"unused", "UnusedAssignment"})
public class TimePickerDialog extends LightBulbPickerDialogFragment<int[]> {
    private static final String TIME_SELECTION_TAG = "TIME_SELECTION_TAG";
    private static final String TIME_SELECTION_DRAFT_TAG = "TIME_SELECTION_DRAFT_TAG";
    private TimePicker picker;
    private boolean ignorePickerEvent;

    public static TimePickerDialog newInstance(
            DialogButtonConfiguration positive, DialogButtonConfiguration negative, boolean cancelable, DialogAnimationTypes animationType) {
        TimePickerDialog f = new TimePickerDialog();
        DialogBundleHelper bundleHelper = new DialogBundleHelper()
                .withPositiveButtonConfig(positive)
                .withNegativeButtonConfig(negative)
                .withCancelable(cancelable)
                .withShowing(false)
                .withDialogType(DialogTypes.NORMAL)
                .withAnimation(animationType == null ? DialogAnimationTypes.NO_ANIMATION : animationType);
        f.setArguments(bundleHelper.getBundle());
        return f;
    }

    public TimePickerDialog() {
        super(new TimeSelection(new int[]{DateUtils.getHourOfDay(DateUtils.now()), DateUtils.getMinuteOfHour(DateUtils.now())}, null), false);
    }

    @Override
    protected void create(Bundle dialogArguments, Bundle savedInstanceState) {
        if (savedInstanceState == null) {
            if (dialogArguments == null) {
                throw new IllegalArgumentException("Bundle args required");
            }
            if (hasSelection())
                selection.setCurrentSelection(selection.getCurrentSelection());
            else
                selection.setCurrentSelection(dialogArguments.getIntArray(TIME_SELECTION_TAG));
        } else {
            selection.setCurrentSelection(savedInstanceState.getIntArray(TIME_SELECTION_TAG), false);
            selection.setDraftSelection(savedInstanceState.getIntArray(TIME_SELECTION_DRAFT_TAG), false);
        }
    }

    @Override
    protected void saveInstanceState(Bundle outState) {
        if (selection.getCurrentSelection() != null)
            outState.putIntArray(TIME_SELECTION_TAG, selection.getCurrentSelection());
        if (selection.getDraftSelection() != null)
            outState.putIntArray(TIME_SELECTION_DRAFT_TAG, selection.getDraftSelection());
    }

    @Override
    protected View setDialogLayout(LayoutInflater inflater) {
        return View.inflate(getContext(), R.layout.dialog_picker_time_picker, null);
    }

    @Override
    protected void configureContent(View view, Bundle savedInstanceState) {
        picker = view.findViewById(R.id.dialogTimePicker);
        picker.setIs24HourView(true);
        picker.setSaveEnabled(false);
        synchronizeSelectUi();
        picker.setOnTimeChangedListener((timePicker, hourOfDay, minutesOfHour) -> {
            int[] newSelection = new int[]{hourOfDay, minutesOfHour};
            if (isDialogShown())
                selection.setDraftSelection(newSelection);
            else
                selection.setCurrentSelection(newSelection);
        });
    }

    public void setSelection(int hour, int minutes) {
        int newHour = validateHour(hour);
        int newMinute = validateMinute(minutes);
        selection.setCurrentSelection(new int[]{newHour, newMinute});
    }

    @SuppressWarnings("deprecation")
    @Override
    protected void synchronizeSelectUi() {
        int[] newDate = selection.hasDraftSelection() ? selection.getDraftSelection() : selection.getCurrentSelection();
        if (newDate != null) {
            if (picker == null)
                return;
            int hour = newDate[0];
            int minutes = newDate[1];
            int apiLevel = Build.VERSION.SDK_INT;
            if (apiLevel < 23) {
                picker.setCurrentHour(hour);
                picker.setCurrentMinute(minutes);
            } else {
                picker.setHour(hour);
                picker.setMinute(minutes);
            }
        }
    }

    public void setSelection(int[] newSelection) {
        if (newSelection == null)
            selection.setCurrentSelection(null);
        else
            setSelection(newSelection[0], newSelection[1]);
    }

    public void setSelectionFromDate(Date newSelection) {
        if (newSelection == null) setSelection(null);
        else
            setSelection(DateUtils.extractYearFromDate(newSelection), DateUtils.extractMonthOfYearFromDate(newSelection));
    }

    private int validateHour(int hour) {
        int minutes = selection.getCurrentSelection()[1];
        if (hour >= 24) {
            hour = 23;
            minutes = 59;
        }
        if (hour < 0) {
            hour = 0;
        }
        return hour;
    }

    private int validateMinute(int minutes) {
        int hour = selection.getCurrentSelection()[0];
        if (minutes == 60) {
            validateHour(hour + 1);
            minutes = 0;
        }
        if (minutes < 0)
            minutes = 0;
        return minutes;
    }


    private int[] getTimeFromDate(Date date) {
        return date == null ? null : new int[]{DateUtils.getHourOfDay(date), DateUtils.getMinuteOfHour(date)};
    }
}

File: lightbulb-dialogs/src/main/java/com/github/rooneyandshadows/lightbulb/dialogs/picker_dialog_time/TimePickerDialog.kt
package com.github.rooneyandshadows.lightbulb.dialogs.picker_dialog_time

import android.os.Build
import android.os.Bundle
import android.os.Parcel
import android.os.Parcelable
import android.view.LayoutInflater
import android.view.View
import android.widget.TimePicker
import com.github.rooneyandshadows.lightbulb.dialogs.base.BasePickerDialogFragment
import com.github.rooneyandshadows.java.commons.date.DateUtilsOffsetDate
import com.github.rooneyandshadows.lightbulb.dialogs.R
import com.github.rooneyandshadows.lightbulb.dialogs.base.internal.DialogTypes
import com.github.rooneyandshadows.lightbulb.dialogs.picker_dialog_time.TimePickerDialog.*
import java.time.OffsetDateTime

@Suppress("unused", "UNUSED_PARAMETER", "DEPRECATION")
class TimePickerDialog : BasePickerDialogFragment<Time>(
    TimeSelection(
        Time(
            DateUtilsOffsetDate.getHourOfDay(DateUtilsOffsetDate.nowLocal()),
            DateUtilsOffsetDate.getMinuteOfHour(DateUtilsOffsetDate.nowLocal())
        ),
        null
    ), false
) {
    private lateinit var picker: TimePicker
    override var dialogType: DialogTypes
        get() = DialogTypes.NORMAL
        set(value) {}

    companion object {
        fun newInstance(): TimePickerDialog {
            return TimePickerDialog()
        }
    }

    @Override
    override fun doOnCreate(dialogArguments: Bundle?, savedInstanceState: Bundle?) {
        if (savedInstanceState != null) return
        if (hasSelection()) dialogSelection.setCurrentSelection(dialogSelection.getCurrentSelection())
    }

    @Override
    override fun getDialogLayout(layoutInflater: LayoutInflater): View {
        return layoutInflater.inflate(R.layout.dialog_picker_time_picker, null)
    }

    @Override
    override fun setupDialogContent(view: View, savedInstanceState: Bundle?) {
        picker = view.findViewById(R.id.dialogTimePicker)
        picker.setIs24HourView(true)
        picker.isSaveEnabled = false
        dialogSelection.getActiveSelection()?.apply {
            picker.hour = hour
            picker.minute = minute
        }
    }

    @Override
    override fun onSelectionChange(newSelection: Time?) {
        if (!checkIfTimePickerNeedsSync(newSelection)) return
        picker.apply {
            newSelection?.apply {
                val hour = hour
                val minutes = minute
                val apiLevel = Build.VERSION.SDK_INT
                if (apiLevel < Build.VERSION_CODES.M) {
                    picker.currentHour = hour
                    picker.currentMinute = minutes
                } else {
                    picker.hour = hour
                    picker.minute = minutes
                }
            }
        }
    }

    private fun checkIfTimePickerNeedsSync(currentValue: Time?): Boolean {
        val timePickerValue = picker.let {
            Time(it.hour, it.minute)
        }
        return !timePickerValue.compare(currentValue)
    }

    @Override
    override fun doOnViewStateRestored(savedInstanceState: Bundle?) {
        super.doOnViewStateRestored(savedInstanceState)
        picker.setOnTimeChangedListener { _: TimePicker?, hourOfDay: Int, minutesOfHour: Int ->
            val newSelection = Time(hourOfDay, minutesOfHour)
            if (isDialogShown) dialogSelection.setDraftSelection(newSelection)
            else dialogSelection.setCurrentSelection(newSelection)
        }
    }

    @Override
    override fun setSelection(newSelection: Time?) {
        dialogSelection.apply {
            if (newSelection == null) setCurrentSelection(null)
            else setCurrentSelection(newSelection)
        }
    }

    fun setSelection(hour: Int, minutes: Int) {
        dialogSelection.apply {
            val time = Time(hour, minutes)
            setCurrentSelection(time)
        }
    }

    fun setSelection(newSelection: OffsetDateTime?) {
        dialogSelection.apply {
            if (newSelection == null) setCurrentSelection(null)
            else setCurrentSelection(Time.fromDate(newSelection))
        }
    }

    class Time(hour: Int, minute: Int) : Parcelable {
        var hour: Int = -1
            private set
        var minute: Int = -1
            private set

        constructor(parcel: Parcel) : this(
            parcel.readInt(),
            parcel.readInt()
        )

        init {
            val validatedTime = validateTime(hour, minute)
            this.hour = validatedTime[0]
            this.minute = validatedTime[1]
        }

        fun compare(hour: Int, minute: Int): Boolean {
            return hour == this.hour && minute == this.minute
        }

        fun compare(target: Time?): Boolean {
            if (target == null) return false
            return hour == target.hour && minute == target.minute
        }

        fun toArray(): IntArray {
            return intArrayOf(hour, minute)
        }

        private fun validateTime(hour: Int, minute: Int): IntArray {
            val validatedHour = when {
                hour < 0 -> 0
                hour > 23 -> 23
                else -> hour
            }
            val validatedMinute = when {
                minute < 0 -> 0
                minute > 59 -> 59
                else -> minute
            }
            return intArrayOf(validatedHour, validatedMinute)
        }

        override fun writeToParcel(parcel: Parcel, flags: Int) {
            parcel.writeInt(hour)
            parcel.writeInt(minute)
        }

        override fun describeContents(): Int {
            return 0
        }

        companion object CREATOR : Parcelable.Creator<Time> {
            override fun createFromParcel(parcel: Parcel): Time {
                return Time(parcel)
            }

            override fun newArray(size: Int): Array<Time?> {
                return arrayOfNulls(size)
            }

            @JvmStatic
            fun fromDate(date: OffsetDateTime): Time {
                val hour = DateUtilsOffsetDate.getHourOfDay(date)
                val minute = DateUtilsOffsetDate.getMinuteOfHour(date)
                return Time(hour, minute)
            }

            @JvmStatic
            fun fromDateString(dateString: String): Time {
                val date = DateUtilsOffsetDate.getDateFromString(DateUtilsOffsetDate.defaultFormatWithTimeZone, dateString)
                return fromDate(date)
            }
        }
    }
}